ASoC: SOF: ipc4-pcm: Look for best matching hw_config for SSP

ujfalusi opened the pull request:
Instead of just looking for a hw_config with matching rate only it sounds better to try to find the best matching configuration.

If we have multiple hw_configurations with the same rate, but each with different format for example then we have been picking the first config with the matching rate, which can be a problem and it wil depend on how the configs are ordered.

Instead we should be trying to find the best match out of the configs
1. rate + format + channels are matching
2. rate + format are matching
3. rate matching

ujfalusi committed:
ASoC: SOF: ipc4-pcm: Look for best matching hw_config for SSP

Instead of just looking for a hw_config with matching rate only it sounds
better to try to find the best matching configuration.

If we have multiple hw_configurations with the same rate, but each with
different format for example then we have been picking the first config
with the matching rate, which can be a problem and it wil depend on how
the configs are ordered.

Instead we should be trying to find the best match out of the configs
1. rate + format + channels are matching
2. rate + format are matching
3. rate matching

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

singalsu commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -575,17 +577,48 @@ static void ipc4_ssp_dai_config_pcm_params_match(struct snd_sof_dev *sdev, const
+			/* best match found */
 			break;
+		} else if (matching_params < 2 &&
+			   params_rate(params) == le32_to_cpu(hw_config->fsync_rate) &&
```
Change second level to rate and channels and remove third level? Fail if first or second level match is not found. Copier can convert sample format but not rate and channels count.

ujfalusi replied:
let me think about it a bit, this function was not able to fail before, even if it has not found the config.

jsarha replied:
There could of course be some simple ranking about the formats if exact match is not found to prefer higher precision over lower, but in practice (only having s32 format available) this hardly matters.

ujfalusi replied:
I think with this we will find the exact match and if no exact config available then we pick the first one which have matching rate and channels. It can be up to debate which non exact match is the best one, but I feel that if we don't have the exact match available then we are already in a 'this is not going to end up right' space..

ujfalusi commented:
Changes since v1:
- only two levels of match: full (rate, channels, format) and partial (rate, channels)
- fail the match function if neither full or partial match is found.
